### ISSUE TITLE
Can't find a way to disable WebAssembly parser

### ISSUE THREAD

**Issue by NeekSandhu** [webpack/webpack#7264]
# Bug report

**What is the current behavior?**

WebAssembly parser kicks in even when `module.noParse` is set to `/.wasm$/` 

**If the current behavior is a bug, please provide the steps to reproduce.**

Please see this conversation for a deeper context - https://github.com/NeekSandhu/onigasm/issues/2#issuecomment-387949215

```sh
mkdir wasm-test
cd wasm-test
npm init -y
npm i onigasm
npm i -D webpack webpack-cli
```

```javascript
// index.js
import {
    loadWASM
} from 'onigasm'

(async () => {
    // see https://www.npmjs.com/package/onigasm#light-it-up
    await loadWASM(require('file-loader!onigasm/lib/onigasm.wasm'))
})()
```

```javascript
// webpack.config.js

module.exports = {
    mode: 'development',
    entry: __dirname + '/index.js',
    output: {
        filename: 'bundle.js'
    },
    module: {
        noParse: /onigasm.wasm/,
        rules: [{
            test: /\.wasm$/,
            use: "file-loader"
        }]
    }
};
```

After the setup above is compiled, and `grep` is done on `bundle.js` for `[fileLoaderHash].wasm` no result will be found because allegedly, WebAssembly parser interupted the process.

**What is the expected behavior?**
Call to `loadWASM` should get populated with valid URL to `ongiasm.wasm` file. Courtesy of `file-loader`.

**Other relevant information:**
webpack version:  4.8.1
Node.js version: 8.9.1
Operating System: Windows 10 64bit
Additional tools: NA

**Comment by montogeek** [comment]
@xtuc Any insight in this issue?

**Comment by xtuc** [comment]
Not really, this should be handled by Webpack's rules.

I'll look into it, maybe I missed an option.

**Comment by xtuc** [comment]
Also i'm wondering, why can't you use the builtin wasm support in Webpack?

**Comment by NeekSandhu** [comment]
I would love to use Webpack for that but for [certain reasons](https://github.com/NeekSandhu/onigasm/issues/2#issuecomment-387950702), I can't.

In a nushell, `onigasm.wasm` is compiled with emscription and we must have access to WASM runtime memory on JS side through `malloc`, `free`, `HEAP8`, `HEAP32`, etc.

Webpack won't provide those hooks.

Perhaps we should consider emitting something similar to emscription's glue code. That would be phenomenal.

**Comment by xtuc** [comment]
Ok, here's my suggestion.

We could publish the emscripten JavaScript glue on npm and link it against the wasm module in webpack.

Currently you'll have an error like: module `env` not found and we can use an alias to map `env` to the emscripten glue.

However it's not possible yet since the emscripten glue is not publisheable easily.

**Comment by NeekSandhu** [comment]
Totally agree. 

Would be certainly cool, but AFAIK the main challenge here that `emscripten` takes a bunch of compiler flags when compiling and those flags get baked into the glue code. Those flags go hand-in-hand with `.wasm` file and respective `.js` file. The challenge?? How do we go about doing that process without involving `emscripten`?

Well, the flags that actually get baked into `.js` glue are things like memory size, whether or not dynamic memory size increase is allowed, should host kill the WASM runtime, etc. At the moment, these things are passed to the `emcc` command on the CLI and there is no way for us (webpack) to know what those `args` were so we can synthesize the glue code ourselves. Uncool.

But maybe we can brew a new way of declaring those "things". Why not declare those things in a file, lets say `[lib_name].wasm.config` or `[lib_name].wasm.json` that lives in the same directory as the `[lib_name].wasm` itself.

Example:
```json
{
  "EXPORT_NAME": "Onigasm",
  "WASM": 1,
  "NO_EXIT_RUNTIME": 1,
  "NO_FILESYSTEM": 1,
  "TOTAL_MEMORY": 157286400,
  "ALLOW_MEMORY_GROWTH": 1,
  "DEMANGLE_SUPPORT": 1,
  "EXTRA_EXPORTED_RUNTIME_METHODS": ["ccall"]
}
```

Maybe @kripken can chime in and shed some light on this.

**Comment by NeekSandhu** [comment]
However in the meantime I'd really thankfull if `module.noParse` is patched and WebAssembly parser stops interupting.

**Comment by kripken** [comment]
I apologize as I don't have a deep understanding of webpack, and I'm not sure I follow the discussion here. But in general, emscripten will emit a JS file and a wasm file which the JS will use, and emscripten can emit that JS as a module (the `MODULARIZE` option, and soon, also as an [ES6 module too](https://github.com/kripken/emscripten/pull/6483)). So my first thought is, can we use that JS like you'd use any other module, publish it on npm, etc.? What's different about it from other modules? :)

One problem I can see is that the emscripten JS will try to load the wasm using streaming compilation and other browser APIs, which maybe don't exist in node.js or maybe webpack doesn't expect? And instead we'd prefer the JS load the wasm with a standard `require()` or something like that? If that's an issue then it would be easy for us to add an emscripten option to for that, if we know what webpack would work best with.

**Comment by NeekSandhu** [comment]
Thanks for coming by, Alon. As fair as I understand, what Webpack team (everyone really) is trying to accomplish is ability to *literally* import `.wasm` file and have it work out of the box. No strings attached.

Like so:

```javascript
import('./add.wasm').then(({ add }) => {
    console.log(add(22, 2200))
})
```

And really, code above will work when compiled with Webpack 4. But problem comes with complex modules like `onigasm.wasm` where the JavaScript runtime needs access to `free`, `malloc`, `HEAPU8`, etc. Thing is Webpack doesn't provide those taps ATM. But this shouldn't be hard since these are just function calls and access to few variables. The hard part is baking `emcc` compile flags into glue code that Webpack could generate on its own.

Obviously that's a very ambitious task. I think what you mentioned has much solid ground and ease of use:

> emscripten can emit that JS as a module (the MODULARIZE option, and soon, also as an ES6 module too). So my first thought is, can we use that JS like you'd use any other module, publish it on npm

^^ That right there IS what I'd happily take any day. The only problem?? Webpack import graph will ultimately reach `[lib_name].wasm` file:

 `app.js` *imports*
`node_modules/[lib_name]/glue.js` *imports*
`node_modules/[lib_name]/bin/[lib_name].wasm`

When that happens, Webpack's WebAssembly parser will find `imports` in WASM module that it can't figure out, one of which is `env`. And then it explodes and build fails.

That's the challenge if go the regular ES6 based javascript glue route. The WASM imports that Webpack can't understand/provide.

**Comment by kripken** [comment]
Thanks @NeekSandhu , that explained a lot to me!

Makes sense to want to allow `import('./add.wasm')` directly. I think that is what the emscripten-generated JS should do, just like other JS that is run through webpack. And people that want to use the wasm will import that JS normally as well. So far, I don't think there is anything different in the emscripten case, **except** that emscripten doesn't expect people to load its emitted wasm directly (rather, the wasm is an internal component of the combined JS+wasm that emscripten emits, they work closely together and except for special cases it's not practical to expect another tool to generate just one of the two).

About the `env` problem, it sounds like the emscripten JS needs to declare a module containing the wasm imports (would we need to do that in a separate file from the rest of the emscripten-generated JS?), and generate the wasm with proper imports for that name?

**Comment by NeekSandhu** [comment]
> About the env problem, it sounds like the emscripten JS needs to declare a module containing the wasm imports (would we need to do that in a separate file from the rest of the emscripten-generated JS?), and generate the wasm with proper imports for that name?

In my opinion, I wonder why would Webpack even bother complaining about `env` or anything else really. If a user wants to consume "pure" `wasm`, sure webpack needs to "inspect" and fill in the holes since there won't be emscripton glue accompanying it. But in the other case, if webpack can deem that there **IS** glue code present and available, it can just keep quiet and do a regular import *[1]*.

Maybe we can tell Webpack if glue code is accompanying or not by some special flag/tag. It could be either a property in WASM exports like `__emscripton_module`, similar to how we see `Object.defineProperty(exports, "__esModule", { value: true });` in some JS files.

Or we can express that in the emscripton glue code itself using Webpack's query params. For example:

```javascript
// [lib_name]-glue.js (ES6 emitted by emscripton)

async initWASM() {
    const wasm = await import('./[lib_name].wasm?skip_parse=true')
}

```

If there aren't any impediments in going forward with this approach, either on Webpack side or `emscripten` side, I believe this could be pretty neat solution.

### Footnotes
*[1]* "Reguarl import" as how `file-loader` works. Put the imported wasm in build output directory and treat as an async chunk module and `fetch` when imported during runtime.

**Comment by ooflorent** [comment]
Instead of skipping the parsing you could change the `type` of the `/\.wasm$/` rule to something else. I don’t know if the config validator would accept it but it will disable the wasm pipeline for the matching rule.

**Comment by sokra** [comment]
`type: "javascript/auto"`

**Comment by NeekSandhu** [comment]
Thank you so much, the `type` hack worked.

At this point, this issue can be closed unless you guys want to consider enhancing WebAssembly support when its compiled with emscripten.

Described here: https://github.com/webpack/webpack/issues/7264#issuecomment-388253498

TL;DR - skip parse but still "link" `.wasm` file when it's being imported by emscripten's JS glue code

**Comment by glebmachine** [comment]
@sokra how do i declare type in inline require, something like:
`require(!!arraybuffer-loader?type=text./zxing.wasm)`

**Comment by taxilian** [comment]
The type hack does not work in all cases because it seems to have other side-effects; currently trying to make this work with vue.js / webpack 4 and I can't. I set type and I just end up with "__webpack_public_path__ is not defined" after needing to manually tell it not to use esModule.

Whoever designed the wasm support needs to seriously research the technology better; I think it's a great idea and all, but there is no possible way you'll *ever* be able to support all uses of wasm and by trying to force it to do so you've made it impossible for anyone to use webpack usefully with a wasm file that doesn't fit the expected mold; all you needed was a simple flag that would disable the parsing!

**Comment by mikefarmer01** [comment]
@taxilian 
Hi, I'm in a similar situation.
Trying to load a .js wrapping a wasm library compiled with reference-types in VueJS.
I'm on Webpack 5.

Have you found a way to disable the wasm parsing?

**Comment by taxilian** [comment]
Nope, I finally gave up on it as it wasn't a high enough priority to spend
more than half a day trying to work around such an asinine problem in a
library which should have been fixed years ago.

(sorry, I'm still frustrated about the problem... I'm sure someone felt
they had a good reason and had good intentions, but a feature that breaks
real use cases with no workaround is a bug, not a feature, and it really
frustrates me that the people making the decision seem to disagree with
that)
